src/main/kotlin/io/bazel/kotlin/builder/tasks/jvm/CompilationTask.kt

```diff
@@ -161,191 +198,116 @@ internal fun JvmCompilationTask.kaptArgs(
-          // TODO(bencodes) This should probably be enabled via some KSP options
-          "allWarningsAsErrors" to listOf("false"),
-        )
+    // Add custom plugin options from config
+    if (config.optionsMap.isNotEmpty()) {
+      val customOptions =
+        config.optionsMap.entries
+          .map { (k, v) -> k to listOf(v) }
+          .toTypedArray()
 
-      values.forEach { pair ->
-        pair.second.forEach { value ->
-          flag(pair.first, value)
-        }
+      base64Encode("-P", *customOptions) { enc ->
+        "plugin:${config.pluginId}:apoptions=$enc"
       }
     }
   }
-
-private fun Pair<String, List<String>>.asKeyToCommaList() =
-  first to listOf(second.joinToString(","))
-
-internal fun JvmCompilationTask.runPlugins(
-  context: CompilationTaskContext,
-  plugins: InternalCompilerPlugins,
-  compiler: KotlinToolchain.KotlincInvoker,
-): JvmCompilationTask {
-  if (
-    (
-      inputs.processorsList.isEmpty() &&
-        inputs.stubsPluginClasspathList.isEmpty()
-    ) ||
-    inputs.kotlinSourcesList.isEmpty()
-  ) {
-    return this
-  } else {
-    if (!outputs.generatedKspSrcJar.isNullOrEmpty()) {
-      return runKspPlugin(context, plugins, compiler)
-    } else if (!outputs.generatedClassJar.isNullOrEmpty()) {
-      return runKaptPlugin(context, plugins, compiler)
-    } else {
-      return this
-    }
-  }
 }
 
-private fun JvmCompilationTask.runKaptPlugin(
+/**
+ * Generic annotation processing plugin runner.
+ * Replaces KAPT-specific runKaptPlugin() with a generic version.
+ */
+private fun JvmCompilationTask.runAnnotationProcessingPlugin(
   context: CompilationTaskContext,
-  plugins: InternalCompilerPlugins,
+  config: JvmCompilationTask.Inputs.AnnotationProcessingConfig,
+  pluginJar: String,
   compiler: KotlinToolchain.KotlincInvoker,
 ): JvmCompilationTask {
-  return context.execute("kapt (${inputs.processorsList.joinToString(", ")})") {
-    baseArgs()
-      .plus(
-        plugins(
-          options = inputs.stubsPluginOptionsList.filterNot { o -> o.startsWith(plugins.kapt.id) },
-          classpath = inputs.stubsPluginClasspathList,
-        ),
-      ).plus(
-        kaptArgs(context, plugins, "stubsAndApt"),
-      ).flag("-d", directories.generatedClasses)
-      .values(inputs.kotlinSourcesList)
-      .values(inputs.javaSourcesList)
-      .list()
-      .let { args ->
+  val processorNames = config.processorsList.joinToString(", ")
+
+  return context.execute("${config.pluginId} ($processorNames)") {
+    val args =
+      baseArgs()
+        .plus(
+          plugins(
+            options =
+              inputs.stubsPluginOptionsList.filterNot { o ->
+                o.startsWith(config.pluginId)
+              },
+            // Filter out the annotation processing plugin jar to avoid duplicate -Xplugin
+            // It will be added by annotationProcessingArgs() below
+            classpath = inputs.stubsPluginClasspathList.filterNot { it == pluginJar },
```

Comment:
Having poked around a bit in this -- it really doesn't seem to matter if the plugin jar is in the classpath.

If it does ,it's better to sort that out in starlark.